near-sdk-macros/src/core_impl/info_extractor/mod.rs

```diff
@@ -91,7 +91,88 @@ pub struct Returns {
+    ///  - Contract state is written if it is modifying method
+    ///
+    /// In case of panic, state is not written.
+    ///
+    /// # Example:
+    /// ```ignore
+    /// pub fn foo(&mut self);
+    /// ```
     Default,
+
+    /// Return type is specified. But it does not have any specifics.
+    ///
+    /// When the contract call happens, in addition to the Default:
+    ///  - The return value is serialized and returned
+    ///
+    /// # Example:
+    /// ```ignore
+    /// pub fn foo(&mut self) -> u64;
+    /// ```
     General(Type),
-    HandlesResult(Type),
+
+    /// Return type is Result<OkType, ErrType> and the function is marked with #[handle_result].
+    /// ErrType struct implements near_sdk::FunctionError. (i.e. used with #[derive(near_sdk::FunctionError)])
+    ///
+    /// When the contract call happens, in addition to the General:
+    ///  - In case Result value is Ok, the unwrapped object is returned
+    ///  - In case Result value is Err, panic is called and state is not written.
+    ///
+    /// # Example:
+    /// ```ignore
+    /// #[handle_result]
+    /// pub fn foo(&mut self) -> Result<u64, &'static str>;
+    /// ```
+    HandlesResultExplicit(ExplicitResult),
+
+    /// Return type is Result<OkType, ErrType> and, the function is not marked with #[handle_result] and
+    /// ErrType struct implements near_sdk::ContractErrorTrait (i.e. used with #[near_sdk::contract_error])
+    ///
+    /// When the contract call happens, in addition to General:
+    ///  - In case Result value is Err, panic is called and state is not written.
+    ///
+    /// As soon as ErrType implements ContractErrorTrait, it is returned as a well-defined structure.
+    /// You can see the structure in #[contract_error] documentation.
+    /// If the error struct does not implement ContractErrorTrait, the code should not compile.
+    ///  - In case #[unsafe_persist_on_error] is used on method, panic is not called. And the contract state is written. But the extra <method_name>_error method is generated. And this method is called in a new Promise. This method effectively panics with structured error.
+    ///
+    /// Please note #[unsafe_persist_on_error] is not safe. Imagine the user has a balance of 100 $COIN. Then they spent 20 $COIN. But some error occurs and you can't proceed with the transaction. Then the user will lose 20 $COIN in case you forget to revert the balance.
+    ///
+    /// # Example:
+    /// ```ignore
+    /// #[contract_error]
+    /// pub struct MyError;
+    ///
+    /// // if Ok() is returned, everything ok, otherwise panic with well-structured error
+    /// pub fn foo(&mut self) -> Result<u64, MyError>;
+    /// ```
+    ///
+    /// ```ignore
+    /// // Write state anyway.
+    /// // if Ok() is returned, just return. Otherwise call new Promise which will panic with well-structured error.
+    /// #[unsafe_persist_on_error]
+    /// pub fn foo(&mut self) -> Result<u64, MyError>;
```

Comment:
IMO, this is a very bad design choice: smart-contract developers should be responsible for panicking on some errors and omitting others , while, most importantly, never leave the contract in corrupted state.
Is there any real-world use-case where `[unsafe_persist_on_error]` would actually be used?

Reply:
> Is there any real-world use-case where [unsafe_persist_on_error] would actually be used?

@mitinarseny There were several usecases in Keypom and other linkdrop contracts that wanted to keep track of the action, but also signal with a transaction error to the cross-contract or transaction calls.

> IMO, this is a very bad design choice: smart-contract developers should be responsible for panicking on some errors and omitting others , while, most importantly, never leave the contract in corrupted state.

I don't see where this design contradicts your statement. This PR enables more granular error design instead of the current model where developers usually panic in random places of the code and as such potential errors are not encoded in the ABI at all.